py/picca/delta_extraction/astronomical_objects/pk1d_forest.py

```diff
@@ -238,44 +238,52 @@ def rebin(self):
         AstronomicalObjectError if Forest.wave_solution is not 'lin' or 'log'
         """
         bins, rebin_ivar, orig_ivar, w1, w2 = super().rebin()
-        if len(rebin_ivar) == 0:
+        if len(rebin_ivar) == 0 or np.sum(w2) == 0:
             self.exposures_diff = np.array([])
             self.reso = np.array([])
             self.reso_pix = np.array([])
-            return [], [], [], [], []
+            return [], [], [], [], np.array([])
 
         # apply mask due to cuts in bin
         self.exposures_diff = self.exposures_diff[w1]
         self.reso = self.reso[w1]
         self.reso_pix = self.reso_pix[w1]
 
-        # rebin exposures_diff and reso
-        rebin_exposures_diff = np.zeros(bins.max() + 1)
-        rebin_reso = np.zeros(bins.max() + 1)
-        rebin_reso_pix = np.zeros(bins.max() + 1)
-        rebin_exposures_diff_aux = np.bincount(bins,
-                                               weights=orig_ivar[w1] *
-                                               self.exposures_diff)
-        rebin_reso_aux = np.bincount(bins, weights=orig_ivar[w1] * self.reso)
-        rebin_reso_pix_aux = np.bincount(bins, weights=orig_ivar[w1] * self.reso_pix)
-        rebin_exposures_diff[:len(rebin_exposures_diff_aux
-                                 )] += rebin_exposures_diff_aux
-        rebin_reso[:len(rebin_reso_aux)] += rebin_reso_aux
-        rebin_reso_pix[:len(rebin_reso_pix_aux)] += rebin_reso_pix_aux
+        # Find non-empty bins
+        binned_arr_size = bins.max() + 1
+        bincounts = np.bincount(bins, minlength=binned_arr_size)
+        wnonempty_bins = bincounts != 0
+        final_arr_size = np.sum(wnonempty_bins)
 
+        # rebin exposures_diff and reso
+        rebin_exposures_diff = np.bincount(bins,
+            weights=orig_ivar[w1] * self.exposures_diff, minlength=binned_arr_size)
+        rebin_reso = np.bincount(bins, weights=orig_ivar[w1] * self.reso, minlength=binned_arr_size)
+        rebin_reso_pix = np.bincount(bins, weights=orig_ivar[w1] * self.reso_pix,
+                                     minlength=binned_arr_size)
+
+        # Remove empty bins but not ivar
+        w2_ = (rebin_ivar > 0.) & wnonempty_bins
+        self.exposures_diff = np.zeros(final_arr_size)
+        self.reso = np.zeros(final_arr_size)
+        self.reso_pix = np.zeros(final_arr_size)
 
         # apply mask due to rebinned inverse vairane
-        self.exposures_diff = rebin_exposures_diff[w2] / rebin_ivar[w2]
-        self.reso = rebin_reso[w2] / rebin_ivar[w2]
-        self.reso_pix = rebin_reso_pix[w2] / rebin_ivar[w2]
+        self.exposures_diff[w2] = rebin_exposures_diff[w2_] / rebin_ivar[w2_]
+        self.reso[w2] = rebin_reso[w2_] / rebin_ivar[w2_]
+        self.reso_pix[w2] = rebin_reso_pix[w2_] / rebin_ivar[w2_]
 
         # finally update control variables
-        self.mean_reso = self.reso.mean()
+        self.mean_reso = self.reso[w2].mean()
         self.mean_z = (
```

Comment:
I guess here we'd need to decide if we want to use the mean redshift of pixels that actually make it into the masked spectrum or the central redshift of the spectrum before masking (which is the behaviour now and has also been previously except if the edge of the spectrum has been masked)